# 안녕하세요^^ 
## AIVLE 1차 미니프로젝트 '서울시 생활정보 기반 대중교통 수요 분석 ' 과정에 오신 여러분을 환영합니다.
* 본 과정에서는 실제 사례와 데이터를 기반으로 문제를 해결하는 전체 과정을 자기 주도형 실습으로 진행해볼 예정입니다.
* 앞선 교육과정을 정리하는 마음과 지금까지 배운 내용을 바탕으로 문제 해결을 해볼게요!
* 미니 프로젝트를 통한 문제 해결 과정 'A에서 Z까지', 지금부터 시작합니다!

---

# 데이터 분석부터 먼저 시작해보겠습니다.
## "구 별 유동인구 데이터" 를 확인해 보도록 하겠습니다

In [ ]:
# 필요 라이브러리부터 설치합니다.
%pip install pandas seaborn

Note: you may need to restart the kernel to use updated packages.


## 기본전제
* 처음에 제공되는 기본/추가 데이터는 'raw_data' 폴더에 있습니다.

[기본 데이터]
* seoul_moving_month_4_step1.csv

[데이터 소개]
* 서울 시 구별 이동 2022년 4월 데이터

[변수 소개]
* 도착시간 : 0 -> "0:00~0:59"를 의미 
* 출발 자치구 : 자치구 코드
* 성별 : F(여성), M(남성)
* 나이 : 10 ~ 79세 (5세 단위), 0 ~ 10세, 80세 이상으로 구분
* 유형 : H(야간상주지), W(주간상주지), E(기타)
* 평균 이동 시간 : 동일 열 내 이동인구의 월 단위 평균 이동 시간 (10분단위)
* 이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만의 경우 * 처리

---

# 1.데이터 불러오기
## 모든 미니 프로젝트의 시작은 '데이터 불러오기' 부터라고 할 수 있습니다.
+ KeyPoint : 불러오고자 하는 데이터에 따라 자유롭게 변수로 지정할 수 있다.

###  데이터 프레임을 불러오고 변수로 저장(여기서는 CSV 기준으로 진행)
* csv : pd.read_csv("파일이름. csv")
* txt : pd.read_csv("파일이름. csv", sep="구분자")
* xlsx : pd.read_excel('파일이름.xlsx')
* pickle : pd.read_pickle("파일이름.pkl") <br>
 [참고] pickle은 파이썬의 모든 객체를 파일로 저장할 수 있는 방법으로 DataFrame,List,Dict 등 모든 객체 저장 가능(특히 sklearn라이브러리를 통해 모델을 학습시키고, 저장할 때 많이 사용)

#### [실습문제1] 데이터 로딩
* Pandas 라이브러리를 활용해서 'seoul_moving_month_4_step1.csv' 데이터를 확인하세요.
    * 아래에 'Pandas'를 Import하고 'seoul_moving_month_4_step1.csv'파일을 'seoul_moving'에 불러오고,  Shape을 확인하는 소스코드를 작성하고 실행해보세요.
    * 데이터 파일 로딩시 참고 사항 
        * 구분자(sep)는 ',' 입니다
        * UTF-8 인코더를 사용해 주세요
        * index column은 0번 열 입니다

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
import pandas as pd

seoul_moving = pd.read_csv('1.2 seoul_moving_month_4.csv',sep=",", index_col = 0, encoding = 'UTF-8')

In [ ]:
# 데이터 프레임의 Shape을 확인합니다.
seoul_moving.shape

(28923054, 10)

---

# 2.기본 정보 확인 및 클렌징

+ 데이터 클렌징 : 결측치, 이상치 등을 제거하여 데이터 분석 결과가 왜곡 되는 문제를 방지하기 위한 정제 과정

#### [실습문제2] 기본 정보 확인하기
* 'seoul_moving' 데이터의 정보를 확인해보세요.
* 'describe', 'info', 'head' 등 전부 활용해 보겠습니다.

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 28923054 entries, 0 to 968462
Data columns (total 10 columns):
 #   Column       Dtype 
---  ------       ----- 
 0   대상연월         int64 
 1   요일           object
 2   도착시간         int64 
 3   출발 시군구 코드    int64 
 4   도착 시군구 코드    int64 
 5   성별           object
 6   나이           int64 
 7   이동유형         object
 8   평균 이동 시간(분)  int64 
 9   이동인구(합)      object
dtypes: int64(6), object(4)
memory usage: 2.4+ GB


In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving.describe()


대상연월          도착시간     출발 시군구 코드     도착 시군구 코드            나이  \
count  28923054.0  2.892305e+07  2.892305e+07  2.892305e+07  2.892305e+07   
mean     202204.0  1.365597e+01  1.712362e+04  1.719750e+04  4.279614e+01   
std           0.0  5.849675e+00  9.046478e+03  9.063818e+03  1.777485e+01   
min      202204.0  0.000000e+00  1.101000e+04  1.101000e+04  0.000000e+00   
25%      202204.0  1.000000e+01  1.110000e+04  1.111000e+04  3.000000e+01   
50%      202204.0  1.400000e+01  1.120000e+04  1.120000e+04  4.500000e+01   
75%      202204.0  1.800000e+01  3.101400e+04  3.102100e+04  5.500000e+01   
max      202204.0  2.300000e+01  3.900000e+04  3.900000e+04  8.000000e+01   

        평균 이동 시간(분)  
count  2.892305e+07  
mean   3.279829e+01  
std    2.584213e+01  
min    1.000000e+01  
25%    2.000000e+01  
50%    3.000000e+01  
75%    4.000000e+01  
max    7.700000e+02

In [ ]:
seoul_moving.head()

대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분) 이동인구(합)
0  202204  일     0      11010      11010  F   0   EE           10   28.02
1  202204  일     0      11010      11010  F   0   EH           10   27.52
2  202204  일     0      11010      11010  F   0   HE           10   27.83
3  202204  일     0      11010      11010  F  10   HW           10    6.05
4  202204  일     0      11010      11010  F  10   EH           90    5.95

* 1번 열의 의미
    * " 2022년 1월 일요일에 11010(종로구)에서 11010(종로구)으로 이동하여 0시 ~ 0시 59분 사이에 도착했고, E(기타)에서 H(야간상주지(집))으로 이동한 0 ~ 10세 여성은 총 27.52명이며 이들의 평균 이동시간은 10분이다

In [ ]:
seoul_moving.tail()

대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분) 이동인구(합)
968458  202204  토    23      39000      11250  M  45   EH           50    3.16
968459  202204  토    23      39000      11250  M  50   EH           70    3.27
968460  202204  토    23      39000      11250  M  55   EE           50    3.31
968461  202204  토    23      39000      11250  M  55   EH           50    3.28
968462  202204  토    23      39000      11250  M  75   EH           30    5.94

#### [실습문제3] 결측치 처리 ( NaN이 아닌 결측치를 처리 해봅시다)

* '인구이동(합)' column은 이동 인구의 합산인데 object 타입입니다. 

    - 왜 일까요? 

    - Tip! (이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만의 경우 * 처리)

    - 해당 값을 나름의 기준으로 변경하세요

    - '인구이동(합)' 컬럼의 데이터 타입을 수치형으로 변경하세요 

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

seoul_moving['이동인구(합)'].value_counts()

*          3842041
3.3         399374
6.05        366830
3.29        289731
3.31        232069
            ...   
2352.37          1
1617.99          1
3141.86          1
1485.54          1
979.87           1
Name: 이동인구(합), Length: 135812, dtype: int64

In [ ]:
# 최빈값과 평균값을 알 수 없는 상황이라 중위값으로 대체

seoul_moving.replace("*","1.5", inplace= True)
seoul_moving['이동인구(합)'].value_counts()

1.5        3842041
3.3         399374
6.05        366830
3.29        289731
3.31        232069
            ...   
2352.37          1
1617.99          1
3141.86          1
1485.54          1
979.87           1
Name: 이동인구(합), Length: 135812, dtype: int64

In [ ]:
seoul_moving.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 28923054 entries, 0 to 968462
Data columns (total 10 columns):
 #   Column       Dtype 
---  ------       ----- 
 0   대상연월         int64 
 1   요일           object
 2   도착시간         int64 
 3   출발 시군구 코드    int64 
 4   도착 시군구 코드    int64 
 5   성별           object
 6   나이           int64 
 7   이동유형         object
 8   평균 이동 시간(분)  int64 
 9   이동인구(합)      object
dtypes: int64(6), object(4)
memory usage: 2.4+ GB


In [ ]:
# 이동인구(합) column의 데이터 타입을 바꿔주세요
seoul_moving = seoul_moving.astype({'이동인구(합)':'float'})
seoul_moving.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 28923054 entries, 0 to 968462
Data columns (total 10 columns):
 #   Column       Dtype  
---  ------       -----  
 0   대상연월         int64  
 1   요일           object 
 2   도착시간         int64  
 3   출발 시군구 코드    int64  
 4   도착 시군구 코드    int64  
 5   성별           object 
 6   나이           int64  
 7   이동유형         object 
 8   평균 이동 시간(분)  int64  
 9   이동인구(합)      float64
dtypes: float64(1), int64(6), object(3)
memory usage: 2.4+ GB


#### [실습문제4] 이동유형, 요일, 성별 변수의 value수를 파악

* 역시나 value_counts()를 활용해 봅시다.

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

seoul_moving['이동유형'].value_counts()

EH    5757073
HE    5520005
EE    5085099
WH    3108026
WE    2809148
HW    2695189
EW    2304535
WW    1026134
HH     617845
Name: 이동유형, dtype: int64

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

seoul_moving['요일'].value_counts()

토    4690806
금    4641881
목    4011291
수    3955646
화    3926876
월    3906272
일    3790282
Name: 요일, dtype: int64

* 4월의 요일 수를 알려드릴게요
    * 일 : 4, 월 : 4, 화 : 4, 수 : 4, 목 : 4, 금 : 5, 토 : 5

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

seoul_moving['성별'].value_counts()

M    16355029
F    12568025
Name: 성별, dtype: int64

#### [실습문제5] 어느 요일의 심야 이동이 가장 많을까요?

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

seoul_moving_night = seoul_moving[seoul_moving['도착시간']<=5]
seoul_moving_night['요일'].value_counts()

토    501891
금    435296
일    383611
목    377407
수    371376
화    359741
월    338716
Name: 요일, dtype: int64

#### [실습문제6] 서울 지역의 데이터만 골라 보도록 하겠습니다.

* 우리는 서울시의 데이터만 필요합니다!

* 출발과 도착이 모두 서울인 데이터만 추출해 주세요.

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

seoul_moving = seoul_moving[(seoul_moving["출발 시군구 코드"] < 20000) & (seoul_moving["도착 시군구 코드"] < 20000)]
seoul_moving

대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분)  \
0       202204  일     0      11010      11010  F   0   EE           10   
1       202204  일     0      11010      11010  F   0   EH           10   
2       202204  일     0      11010      11010  F   0   HE           10   
3       202204  일     0      11010      11010  F  10   HW           10   
4       202204  일     0      11010      11010  F  10   EH           90   
...        ... ..   ...        ...        ... ..  ..  ...          ...   
920410  202204  토    23      11250      11250  M  80   WH           10   
920411  202204  토    23      11250      11250  M  80   EH           10   
920412  202204  토    23      11250      11250  M  80   EE           20   
920413  202204  토    23      11250      11250  M  80   HE           10   
920414  202204  토    23      11250      11250  M  80   WE           10   

        이동인구(합)  
0         28.02  
1         27.52  
2         27.83  
3          6.05  
4          5.95  
...         ...  
920410    85.10  
920411    28.36  
920412    37.99  
920413     9.42  
920414     9.51  

[10655895 rows x 10 columns]

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

#### [실습문제7] 서울지역의 구 별로 groupby하여 데이터를 확인하겠습니다 (유동인구 기준)

* 유출이 제일 많은 구는?

* 유입이 제일 많은 구는?

* 유출이 제일 적은 구는?

* 유입이 제일 적은 구는?

* 참고자료 서울 구별 코드
    * 11010	종로구
    * 11020	중구
    * 11030	용산구
    * 11040	성동구
    * 11050	광진구
    * 11060	동대문구
    * 11070	중랑구
    * 11080	성북구
    * 11090	강북구
    * 11100	도봉구
    * 11110	노원구
    * 11120	은평구
    * 11130	서대문구
    * 11140	마포구
    * 11150	양천구
    * 11160	강서구
    * 11170	구로구
    * 11180	금천구
    * 11190	영등포구
    * 11200	동작구
    * 11210	관악구
    * 11220	서초구
    * 11230	강남구
    * 11240	송파구
    * 11250	강동구

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

pd.options.display.float_format = '{:.2f}'.format

seoul_moving.groupby(by=['출발 시군구 코드'], as_index=False)["이동인구(합)"].sum()


출발 시군구 코드     이동인구(합)
0       11010 11067062.84
1       11020 11421970.40
2       11030  8236914.94
3       11040 10170899.73
4       11050 10573297.60
5       11060  9377633.96
6       11070  8351684.49
7       11080 11388919.73
8       11090  7596544.82
9       11100  6446373.46
10      11110 12504441.36
11      11120 10292181.59
12      11130 10309642.65
13      11140 14846647.88
14      11150  8989861.03
15      11160 13973493.57
16      11170  9215336.82
17      11180  5197043.78
18      11190 16148071.54
19      11200 10206030.40
20      11210 11376382.34
21      11220 19010806.56
22      11230 30353759.84
23      11240 23626714.17
24      11250 14482722.16

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving.groupby(by=['도착 시군구 코드'], as_index=False)["이동인구(합)"].sum()


도착 시군구 코드     이동인구(합)
0       11010 11521454.69
1       11020 11526178.91
2       11030  8731066.25
3       11040 10376463.41
4       11050 11014786.16
5       11060  9509691.96
6       11070  8469007.57
7       11080 11270988.10
8       11090  7727739.40
9       11100  6533455.94
10      11110 12530704.48
11      11120 10452293.76
12      11130 10195984.99
13      11140 15076453.22
14      11150  9281800.65
15      11160 14008312.51
16      11170  9166928.80
17      11180  5122012.58
18      11190 15686297.00
19      11200 10111772.37
20      11210 11630903.18
21      11220 18644198.64
22      11230 29182496.02
23      11240 22997744.29
24      11250 14395702.78

#### [실습문제8] 낮시간만 볼게요 심야버스 제외하겠습니다

* 도착시간 0부터 4까지(0:00 ~ 4:59) 제거

In [ ]:
index = seoul_moving[seoul_moving['도착시간'] <=4].index
seoul_moving.drop(index, inplace=True)
seoul_moving.reset_index(drop=True, inplace=True)
seoul_moving

대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분)  \
0        202204  일     5      11110      11080  F  65   EW           10   
1        202204  일     5      11110      11080  F  65   WH           10   
2        202204  일     5      11110      11080  F  70   HW           10   
3        202204  일     5      11110      11080  F  70   HE           10   
4        202204  일     5      11110      11080  F  70   EH           20   
...         ... ..   ...        ...        ... ..  ..  ...          ...   
5489130  202204  토    23      11250      11250  M  80   WH           10   
5489131  202204  토    23      11250      11250  M  80   EH           10   
5489132  202204  토    23      11250      11250  M  80   EE           20   
5489133  202204  토    23      11250      11250  M  80   HE           10   
5489134  202204  토    23      11250      11250  M  80   WE           10   

         이동인구(합)  
0           4.68  
1           9.44  
2           5.69  
3           5.68  
4           5.73  
...          ...  
5489130    85.10  
5489131    28.36  
5489132    37.99  
5489133     9.42  
5489134     9.51  

[5489135 rows x 10 columns]

[실습문제9] '도착 시군구 코드' 를 기준으로 각 구별로 '평균 이동 시간(분)' , '이동인구(합)' 의 평균 과 합을 구해보겠습니다

* groupby 사용, as_index=False로 설정해주세요

In [ ]:
df_mean = seoul_moving.groupby(by=['도착 시군구 코드'], as_index=False)[['평균 이동 시간(분)','이동인구(합)']].mean()
df_mean

도착 시군구 코드  평균 이동 시간(분)  이동인구(합)
0       11010        21.87    46.05
1       11020        21.52    46.21
2       11030        21.82    38.41
3       11040        22.15    44.72
4       11050        22.81    50.04
5       11060        22.15    41.04
6       11070        23.89    43.64
7       11080        22.72    48.93
8       11090        24.36    42.22
9       11100        25.14    40.19
10      11110        24.66    59.03
11      11120        24.99    55.02
12      11130        22.41    46.13
13      11140        22.31    60.52
14      11150        23.21    49.14
15      11160        25.52    65.16
16      11170        23.16    45.01
17      11180        24.28    32.82
18      11190        21.75    61.74
19      11200        22.06    45.12
20      11210        23.48    53.02
21      11220        23.12    72.51
22      11230        23.42   104.44
23      11240        25.74    94.94
24      11250        25.93    73.45

In [ ]:
df_sum = seoul_moving.groupby(by=['도착 시군구 코드'], as_index=False)[['평균 이동 시간(분)','이동인구(합)']].sum()
df_sum.rename(columns={'평균 이동 시간(분)':'총 이동 시간','이동인구(합)':'총 이동인구'},inplace=True)
df_sum

도착 시군구 코드  총 이동 시간      총 이동인구
0       11010  5471640 11521454.69
1       11020  5368440 11526178.91
2       11030  4959790  8731066.25
3       11040  5139830 10376463.41
4       11050  5020720 11014786.16
5       11060  5132710  9509691.96
6       11070  4635870  8469007.57
7       11080  5232500 11270988.10
8       11090  4458260  7727739.40
9       11100  4086140  6533455.94
10      11110  5235480 12530704.48
11      11120  4748160 10452293.76
12      11130  4952990 10195984.99
13      11140  5557360 15076453.22
14      11150  4383860  9281800.65
15      11160  5486630 14008312.51
16      11170  4717560  9166928.80
17      11180  3789660  5122012.58
18      11190  5525180 15686297.00
19      11200  4944030 10111772.37
20      11210  5150450 11630903.18
21      11220  5944440 18644198.64
22      11230  6543820 29182496.02
23      11240  6234190 22997744.29
24      11250  5082380 14395702.78

In [ ]:
# 파일을 합쳐볼까요

# df_mean
# df_sum

df_seoul_moving = pd.merge(df_mean,df_sum,how='inner', on='도착 시군구 코드')
df_seoul_moving

도착 시군구 코드  평균 이동 시간(분)  이동인구(합)  총 이동 시간      총 이동인구
0       11010        21.87    46.05  5471640 11521454.69
1       11020        21.52    46.21  5368440 11526178.91
2       11030        21.82    38.41  4959790  8731066.25
3       11040        22.15    44.72  5139830 10376463.41
4       11050        22.81    50.04  5020720 11014786.16
5       11060        22.15    41.04  5132710  9509691.96
6       11070        23.89    43.64  4635870  8469007.57
7       11080        22.72    48.93  5232500 11270988.10
8       11090        24.36    42.22  4458260  7727739.40
9       11100        25.14    40.19  4086140  6533455.94
10      11110        24.66    59.03  5235480 12530704.48
11      11120        24.99    55.02  4748160 10452293.76
12      11130        22.41    46.13  4952990 10195984.99
13      11140        22.31    60.52  5557360 15076453.22
14      11150        23.21    49.14  4383860  9281800.65
15      11160        25.52    65.16  5486630 14008312.51
16      11170        23.16    45.01  4717560  9166928.80
17      11180        24.28    32.82  3789660  5122012.58
18      11190        21.75    61.74  5525180 15686297.00
19      11200        22.06    45.12  4944030 10111772.37
20      11210        23.48    53.02  5150450 11630903.18
21      11220        23.12    72.51  5944440 18644198.64
22      11230        23.42   104.44  6543820 29182496.02
23      11240        25.74    94.94  6234190 22997744.29
24      11250        25.93    73.45  5082380 14395702.78

In [ ]:
# 해당 데이터프레임을 csv 파일로 저정하세요

df_seoul_moving.to_csv("df_seoul_moving.csv", index = False)

---

# 3.데이터 분석하기
+ KeyPoint : 데이터의 형태를 살펴보고 다양한 분석기법을 통해 모델링에 적합하도록 정제요소를 선별할 수 있다.
  * 데이터들의 패턴 탐색
  * 변수들간의 관계 파악

In [ ]:
# 시각화 한글폰트 설정
import seaborn as sns
import matplotlib.pyplot as plt

plt.rc('font', family='Malgun Gothic')
sns.set(font="Malgun Gothic",#"NanumGothicCoding", 
        rc={"axes.unicode_minus":False}, # 마이너스 부호 깨짐 현상 해결
        style='darkgrid')

---